### PR TITLE
[G-API] Introduce GOpaque and GArray for python

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### Overview
This PR is a part of https://github.com/opencv/opencv/pull/18900. To make review process easier decided to split this on several PR's.    

Templates classes that GOpaque/GArray are can't be wrapped into python by using macros. (Parser ignores templates objects)   

* Added `gapi/include/python_bridge.hpp` and `gapi/src/gapi/python_bridge.cpp`. They must be in public api to allow `python` module work with entities from there.
* Implemented `GOpaqueT`/`GArrayT` wrapper, thus all types collected in a single class. This makes it easier to handle them.
* `ArgType` is implemented as `enum` not `enum class` to use `cv.gapi.CV_BOOL` from python instead of `cv.gapi.ArgType_CV_BOOL`.
* `ArgType` contains `CV_GMAT` value to work with `GArray<cv::GMat>` 

### Build configuration
```
Xforce_builders_only=linux,docs
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

build_image:Custom=centos:7
buildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

Xbuild_image:Custom=ubuntu-openvino-2021.1.0:20.04
build_image:Custom Win=openvino-2021.1.0
build_image:Custom Mac=openvino-2021.1.0

Xtest_modules:Custom=gapi
test_modules:Custom Win=gapi
test_modules:Custom Mac=gapi

Xbuildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
Xtest_opencl:Custom=OFF
Xtest_bigdata:Custom=1
Xtest_filter:Custom=*
```

### CODE REVIEW

**Comment by TolyaTalamanov** [comment]
@dmatveev @smirnov-alexey

**Comment by TolyaTalamanov** [comment]
@dmatveev  Could you take a look ?

File: modules/gapi/include/opencv2/gapi/python_bridge.hpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+                                            , cv::Mat
+                                            , cv::GMat>;
+    template<typename T>
+    GArrayT(cv::GArray<T> arg) : m_arg(arg) { };
+
+    GAPI_WRAP GArrayT(gapi::ArgType type) : m_type(type)
+    {
+
+#define HANDLE_CASE(T, K) case gapi::ArgType::CV_##T: \
+                m_arg = cv::GArray<K>(); \
```
Comment (smirnov-alexey) [inline modules/gapi/include/opencv2/gapi/python_bridge.hpp:176]: Can some of this code become template to support both array and opaque? All those duplicates of `HANDLE_CASE` with the same types looks way too massive

Comment (smirnov-alexey) [reply modules/gapi/include/opencv2/gapi/python_bridge.hpp:176]: E.g. move some switches to template functions which return desired type and then use those functions in constructors/etc

**Review by smirnov-alexey — COMMENTED** [review]
Except for code duplication looks good

File: modules/gapi/include/opencv2/gapi/python_bridge.hpp (hunk 1 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/python_bridge.hpp:176]: What should return this template function ?

**Review by smirnov-alexey — APPROVED** [review]

File: modules/gapi/include/opencv2/gapi/python_bridge.hpp (hunk 1 above)
Comment (smirnov-alexey) [reply modules/gapi/include/opencv2/gapi/python_bridge.hpp:176]: Can try to utilize something like `getFromStream`, non-blocking though

File: modules/gapi/include/opencv2/gapi/python_bridge.hpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            HANDLE_CASE(double);
+            HANDLE_CASE(float);
+            HANDLE_CASE(std::string);
+            HANDLE_CASE(cv::Point);
+            HANDLE_CASE(cv::Point2f);
+            HANDLE_CASE(cv::Size);
+            HANDLE_CASE(cv::Rect);
+            HANDLE_CASE(cv::Scalar);
+            HANDLE_CASE(cv::Mat);
+            HANDLE_CASE(cv::GMat);
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/python_bridge.hpp:216]: Giving you some idea.

```c++
#include <iostream>
#include <tuple>
#include <vector>
#include <string>
#include <type_traits>

#define ID(X)       X
#define ID_(X)		ID(X),
#define NOTHING(X)
#define SPACE
#define COMMA       ,

#define TYPE_LIST_G(G, X, A1, L1, A2, L2) \
	G(A1(int)    X  A2("int")) \
	G(A1(double) X  A2("double")) \
	G(L1(char)   X  L2("char"))
	
#define TYPE_LIST(A1, L1, A2, L2) \
	TYPE_LIST_G(ID, SPACE, A1, L1, A2, L2)

#define TYPE_LIST2_G(G, A1, A2) \
	TYPE_LIST_G(G, COMMA, A1, A1, A2, A2)

int main() {
	// Example 1: generate a tuple from our type list
	// A1 is identity plus ','
	// L1 is identity (no comma)
	// A2 is nothing
	// L2 is nothing
	using T = std::tuple< TYPE_LIST(ID_, ID, NOTHING, NOTHING) >;
	static_assert(std::is_same<int,	   std::tuple_element<0, T>::type>::value, "First type is int");
	static_assert(std::is_same<double, std::tuple_element<1, T>::type>::value, "Second type is double");
	
	// Example 2: generate a string list from our type list
	// A1 is nothing
	// L1 is nothing
	// A2 is identity plus ','
	// L2 is identity (no comma)
	std::vector<std::string> names = {
		TYPE_LIST(NOTHING, NOTHING, ID_, ID)
	};
	for (const auto &s : names) std::cout << s << std::endl;
	
	// Example 3: generate code around elements
	// Functors for all and terminating elements are the same
#define HANDLE_CASE(T, S) std::cout << "sizeof(" << S << ") is " << sizeof(T) << std::endl; 
	TYPE_LIST2_G(HANDLE_CASE, ID, ID)
	return 0;
}
```

```
int
double
char
sizeof(int) is 4
sizeof(double) is 8
sizeof(char) is 1
```

**Review by dmatveev — CHANGES_REQUESTED** [review]

File: modules/gapi/include/opencv2/gapi/python_bridge.hpp [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    CV_DOUBLE,
+    CV_FLOAT,
+    CV_STRING,
+    CV_POINT,
+    CV_POINT2F,
+    CV_SIZE,
+    CV_RECT,
+    CV_SCALAR,
+    CV_MAT,
+    CV_GMAT,
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/python_bridge.hpp:30]: * `It is placed in `cv::detail` - `cv.detail_CV_MAT` from python looks ugly
* `CV_GMAT` isn't in this `enum` you mentioned

File: modules/gapi/include/opencv2/gapi/python_bridge.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/python_bridge.hpp:216]: Many thanks! Fixed

**Comment by TolyaTalamanov** [comment]
@dmatveev Could you check it ?

**Review by dmatveev — APPROVED** [review]
**LGTM**

**Comment by alalek** [comment]
Please take a look on Windows compilation errors (they were here before my commit too)

File: modules/gapi/include/opencv2/gapi/python_bridge.hpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2021 Intel Corporation
+
+#ifndef OPENCV_GAPI_PYTHON_BRIDGE_HPP
+#define OPENCV_GAPI_PYTHON_BRIDGE_HPP
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/python_bridge.hpp]: Agree, but didn't know how to do that. Thank for you commit.

File: modules/gapi/include/opencv2/gapi/python_bridge.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/python_bridge.hpp]: Will file a ticket, thanks!

**Pull request merged**